[AOTAutogradCache] Allow Custom Autograd functions behind a flag

jamesjwu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #149751

This adds a new env var and flag,

autograd_cache_allow_custom_autograd_functions, (env var: `TORCHINDUCTOR_AUTOGRAD_CACHE_ALLOW_CUSTOM_AUTOGRAD`) which allows custom autograd functions into AOTAutogradCache.

@hirsheybar and I worked together to verify that the higher order op AutogradFunctionApply is pure with respect to the dynamo input being passed in, so this *should* be safe. I'm still putting it behind a flag and turning it on slowly, first on an internal model, though. Once we verify that it is correct on the internal model we can work to enable the flag by default. 

Differential Revision: [D71633184](https://our.internmc.facebook.com/intern/diff/D71633184/)

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

jamesjwu committed:
Allow custom autograd functions behind a flag

This adds a new env var and flag,

autograd_cache_allow_custom_autograd_functions, (env var: `TORCHINDUCTOR_AUTOGRAD_CACHE_ALLOW_CUSTOM_AUTOGRAD`) which allows custom autograd functions into AOTAutogradCache.

@hirsheybar and I worked together to verify that the higher order op AutogradFunctionApply is pure with respect to the dynamo input being passed in.

Differential Revision: [D71633184](https://our.internmc.facebook.com/intern/diff/D71633184/)

[ghstack-poisoned]

bdhirsh commented on torch/_functorch/_aot_autograd/autograd_cache.py:
```diff
@@ -138,6 +144,12 @@ def is_public_torch_api(target):
         """Allowlisted torch functions"""
         function_name = f"{target.__module__}.{target.__name__}"
+
+        if is_autograd_function(function_name):
```
so existing hops have a `cacheable=True|False` boolean that you can flip, which I imagine is used by inductor's FXGraphCache. Can we just re-use the same infra here? (the main difference is just that `autograd.Function` is a HOP that exists in the dynamo graph, but "disappears" by the time we get to inductor)

bdhirsh replied:
example: https://github.com/pytorch/pytorch/blob/main/torch/_higher_order_ops/triton_kernel_wrap.py#L773

jamesjwu replied:
Oh yeah, we could do that as well, to make this code more generic. We'll still need to add FunctionCtx to the list of safe torch functions, though, since that isn't actually a higher order op

jamesjwu committed:
Update on "Allow custom autograd functions behind a flag"


This adds a new env var and flag,

autograd_cache_allow_custom_autograd_functions, (env var: `TORCHINDUCTOR_AUTOGRAD_CACHE_ALLOW_CUSTOM_AUTOGRAD`) which allows custom autograd functions into AOTAutogradCache.

hirsheybar and I worked together to verify that the higher order op AutogradFunctionApply is pure with respect to the dynamo input being passed in.

Differential Revision: [D71633184](https://our.internmc.facebook.com/intern/diff/D71633184/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

jamesjwu commented:
Updated to use higher order op check instead

jamesjwu committed:
Update on "[AOTAutogradCache] Allow Custom Autograd functions behind a flag"


This adds a new env var and flag,

autograd_cache_allow_custom_autograd_functions, (env var: `TORCHINDUCTOR_AUTOGRAD_CACHE_ALLOW_CUSTOM_AUTOGRAD`) which allows custom autograd functions into AOTAutogradCache.

hirsheybar and I worked together to verify that the higher order op AutogradFunctionApply is pure with respect to the dynamo input being passed in, so this *should* be safe. I'm still putting it behind a flag and turning it on slowly, first on an internal model, though. Once we verify that it is correct on the internal model we can work to enable the flag by default. 

Differential Revision: [D71633184](https://our.internmc.facebook.com/intern/diff/D71633184/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

jamesjwu commented on torch/_functorch/_aot_autograd/autograd_cache.py:
```diff
@@ -138,6 +138,12 @@ def is_public_torch_api(target):
         """Allowlisted torch functions"""
         function_name = f"{target.__module__}.{target.__name__}"
+        # Allow torch.autograd.function.FunctionCtx if custom autograd functions are allowed
+        if function_name == "torch.autograd.function.FunctionCtx":
```
I could have made this unconditionally a safe torch function and relied on line 163 to catch the corresponding autograd higher order op, but I think having it be handled separately is a bit more clear in terms of code reading

bdhirsh commented on torch/_functorch/_aot_autograd/autograd_cache.py:
```diff
@@ -151,6 +157,10 @@ def is_torch_function(target):
         if is_public_torch_api(target):
             return True
+        # Technically, FXGraphCache._check_for_hop already checks this,
+        # but better to error earlier anyway
```
just curious - what extra work happens between here and the eventual call to `FXGraphCache._check_for_hop`?

jamesjwu replied:
Not much, just going through the entire dynamo graph and checking for other ops.

bdhirsh reviewed (approved):
Flipping via config for now sounds ok to me

zou3519 reviewed (approved):
SGTM, but the autograd.Function ctx object shouldn't be in the Dynamo graph. Could you file an issue around that if it's necessary for the caching?

jamesjwu commented on torch/_functorch/_aot_autograd/autograd_cache.py:
```diff
@@ -138,6 +138,12 @@ def is_public_torch_api(target):
         """Allowlisted torch functions"""
         function_name = f"{target.__module__}.{target.__name__}"
+        # Allow torch.autograd.function.FunctionCtx if custom autograd functions are allowed
+        if function_name == "torch.autograd.function.FunctionCtx":
```
This field definitely appears today in the dynamo graph, but gets set to None immediately (as far as I can tell). I can file an issue around it. 

Caching doesn't care whether it appears(in fact, it would be nice if it didn't), but it's in the dynamo graph at least today haha

ad8e commented:
This flag is helpful (10% reduction in compile time). Is it safe now to flip on by default?
